gramineproject/gramine
# [LibOS] Add parser for flock() syscall

## Description of the changes 

This is a preparatory commit to introduce flock (BSD locks) support.

Extracted from #1212. (The initial request from @mkow was to extract only the parser part, but the parser won't even be called if there is no syscall implementation, so I added a dummy one.)

Some references for flags in Linux:
- https://elixir.bootlin.com/linux/latest/source/fs/locks.c#L2061
- https://elixir.bootlin.com/linux/latest/source/include/uapi/asm-generic/fcntl.h#L187
- https://elixir.bootlin.com/linux/latest/source/fs/locks.c#L416

## How to test this PR? 

There is not yet any test for `flock`. But manually can check with this git diff:
```diff
diff --git a/libos/test/regression/helloworld.c b/libos/test/regression/helloworld.c
@@ -1,6 +1,15 @@
 #include <stdio.h>

+#include <sys/file.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#define LOCK_MAND   32
+
 int main(int argc, char** argv) {
     printf("Hello world!\n");
+
+    int fd = open("unknown_testfile", O_RDONLY);
+    flock(fd, LOCK_SH | LOCK_NB | LOCK_MAND | 1024);
     return 0;
 }
```

This will print:
```
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- flock(3, LOCK_SH|LOCK_NB|LOCK_MAND|0x400) ...
(libos_fcntl.c:288:libos_syscall_flock) [P1:T1:helloworld] warning: flock request with LOCK_MAND is ignored
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- return from flock(...) = 0x0
```


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gramine/1404)
